zerver/lib/user_groups.py

```diff
@@ -219,17 +226,96 @@ def access_user_group_for_setting(
             _("'{setting_name}' setting cannot be set to '{group_name}' group.").format(
-                setting_name=setting_name, group_name=user_group.name
+                setting_name=setting_name, group_name=setting_group.name
             )
         )
 
+
+def update_or_create_user_group_for_setting(
+    realm: Realm,
+    direct_members: List[int],
+    direct_subgroups: List[int],
+    current_setting_value: Optional[UserGroup],
+) -> UserGroup:  # nocoverage
+    if current_setting_value is not None and not hasattr(current_setting_value, "named_user_group"):
+        # We do not create a new group if the setting was already set
+        # to an anonymous group. The memberships of existing group
+        # itself are updated.
+        user_group = current_setting_value
+    else:
+        user_group = UserGroup.objects.create(realm=realm)
+
+    from zerver.lib.users import user_ids_to_users
+
+    member_users = user_ids_to_users(direct_members, realm)
+    user_group.direct_members.set(member_users)
+
+    potential_subgroups = NamedUserGroup.objects.filter(realm=realm, id__in=direct_subgroups)
+    group_ids_found = [group.id for group in potential_subgroups]
+    group_ids_not_found = [
+        group_id for group_id in direct_subgroups if group_id not in group_ids_found
+    ]
+    if group_ids_not_found:
+        raise JsonableError(
+            _("Invalid user group ID: {group_id}").format(group_id=group_ids_not_found[0])
+        )
+
+    user_group.direct_subgroups.set(group_ids_found)
+
     return user_group
 
 
+def access_user_group_for_setting(
+    setting_user_group: Union[int, AnonymousSettingGroupDict],
+    user_profile: UserProfile,
+    *,
+    setting_name: str,
+    permission_configuration: GroupPermissionSetting,
+    current_setting_value: Optional[UserGroup] = None,
+) -> UserGroup:
+    if isinstance(setting_user_group, int):
+        named_user_group = access_user_group_by_id(setting_user_group, user_profile, for_read=True)
+        check_setting_configuration_for_system_groups(
+            named_user_group, setting_name, permission_configuration
+        )
+        return named_user_group.usergroup_ptr
+
+    assert "direct_members" in setting_user_group
+    assert "direct_subgroups" in setting_user_group
+
+    if (
+        len(setting_user_group["direct_members"]) == 0
+        and len(setting_user_group["direct_subgroups"]) == 1
+    ):  # nocoverage
+        # This means the setting would ultimately be set to one user group,
+        # so we need not create a new UserGroup and directly use the existing
+        # named user group.
+        setting_group_id = setting_user_group["direct_subgroups"][0]
+        named_user_group = access_user_group_by_id(setting_group_id, user_profile, for_read=True)
+        check_setting_configuration_for_system_groups(
+            named_user_group, setting_name, permission_configuration
+        )
+        return named_user_group.usergroup_ptr
```

Comment:
It would be nice to deduplicate this with the block at the top of the file; I think it might feel easier to do that after resolving https://github.com/zulip/zulip/pull/29937/commits/7df7a3e91b9243c2de6679b8cec21014b96914c1#r1589801457 though.